transport/src/main/java/io/netty/channel/pool/FixedChannelPool.java

```diff
@@ -0,0 +1,362 @@
+
+                    if (future.isSuccess()) {
+                        Boolean result = future.getNow();
+                        if (result == Boolean.TRUE) {
+                            --acquiredChannelCount;
+
+                            // We should never have a negative value.
+                            assert acquiredChannelCount >= 0;
+
+                            // Run the pending acquire tasks before notify the original promise so if the user would
+                            // try to acquire again from the ChannelFutureListener and the pendingAcquireCount is >=
+                            // maxPendingAcquires we may be able to run some pending tasks first and so allow to add
+                            // more.
+                            runTaskQueue();
+                        }
+                        promise.setSuccess(result);
+                    } else {
+                        promise.setFailure(future.cause());
+                    }
+                }
+            }));
+            return p;
+        } catch (Throwable cause) {
+            promise.setFailure(cause);
+        }
+        return promise;
+    }
+
+    private void runTaskQueue() {
+        while (acquiredChannelCount <= maxConnections) {
+            AcquireTask  task = pendingAcquireQueue.poll();
+            if (task == null) {
+                break;
+            }
+
+            // Cancel the timeout if one was scheduled
+            ScheduledFuture<?> timeoutFuture = task.timeoutFuture;
+            if (timeoutFuture != null) {
+                timeoutFuture.cancel(false);
+            }
+
+            --pendingAcquireCount;
+            ++acquiredChannelCount;
+
+            super.acquire(task.key, task.promise);
+        }
+
+        // We should never have a negative value.
+        assert pendingAcquireCount >= 0;
+        assert acquiredChannelCount >= 0;
+    }
+
+    // AcquireTask extends AcquireListener to reduce object creations and so GC pressure
+    private final class AcquireTask extends AcquireListener {
+        final K key;
+        final Promise<C> promise;
+        final long creationTime = System.nanoTime();
+        ScheduledFuture<?> timeoutFuture;
+
+        public AcquireTask(final K key, Promise<C> promise) {
+            super(promise);
+            this.key = key;
+            // We need to create a new promise as we need to ensure the AcquireListener runs in the correct
+            // EventLoop.
+            this.promise = executor.<C>newPromise().addListener(this);
+        }
+    }
+
+    private abstract class TimeoutTask implements Runnable {
+        @Override
+        public final void run() {
+            assert executor.inEventLoop();
+
+            long expiredThresholdTime = System.nanoTime() - acquireTimeoutNanos;
+            for (;;) {
+                AcquireTask task = pendingAcquireQueue.peek();
+                if (task == null || task.creationTime >= expiredThresholdTime) {
```

Comment:
This won't work when the nanoTime overflows.  System.nanoTime() should be used via subtraction only.  Alternatively, you can use the trick we used at [`ScheduledFutureTask.nanoTime()`](https://github.com/netty/netty/blob/master/common/src/main/java/io/netty/util/concurrent/ScheduledFutureTask.java#L34)